File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/TimerHandler.java
package closer.vlllage.com.closer.handler.helpers;


import android.os.Handler;
import android.os.Looper;

import closer.vlllage.com.closer.pool.PoolMember;
import io.reactivex.disposables.Disposable;

public class TimerHandler extends PoolMember {

    private Handler handler;

    @Override
    protected void onPoolInit() {
        handler = new Handler(Looper.getMainLooper());
    }

    public void post(Runnable runnable) {
        handler.post(runnable);
    }

    public void postDisposable(Runnable runnable, long millis) {
        $(DisposableHandler.class).add(post(runnable, millis));
    }

    private Disposable post(Runnable runnable, long millis) {
        DisposableRunnable disposableRunnable = new DisposableRunnable(runnable);
        handler.postDelayed(disposableRunnable, millis);
        return disposableRunnable;
    }

    public class DisposableRunnable implements Disposable, Runnable {

        private final Runnable runnable;
        private boolean disposed;

        private DisposableRunnable(Runnable runnable) {
            this.runnable = runnable;
        }

        @Override
        public void dispose() {
            handler.removeCallbacks(runnable);
            disposed = true;
        }

        @Override
        public boolean isDisposed() {
            return disposed;
        }

        @Override
        public void run() {
            disposed = true;
            runnable.run();
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/TimerHandler.kt
package closer.vlllage.com.closer.handler.helpers


import android.os.Handler
import android.os.Looper

import com.queatz.on.On
import com.queatz.on.OnLifecycle
import io.reactivex.disposables.Disposable

class TimerHandler constructor(private val on: On) : OnLifecycle {

    private lateinit var handler: Handler

    override fun on() {
        handler = Handler(Looper.getMainLooper())
    }

    fun post(runnable: Runnable) {
        handler.post(runnable)
    }

    fun postDisposable(runnable: Runnable, millis: Long) {
        on<DisposableHandler>().add(post(runnable, millis))
    }

    private fun post(runnable: Runnable, millis: Long): Disposable {
        val disposableRunnable = DisposableRunnable(runnable)
        handler.postDelayed(disposableRunnable, millis)
        return disposableRunnable
    }

    inner class DisposableRunnable constructor(private val runnable: Runnable) : Disposable, Runnable {
        private var disposed: Boolean = false

        override fun dispose() {
            handler.removeCallbacks(runnable)
            disposed = true
        }

        override fun isDisposed() = disposed

        override fun run() {
            disposed = true
            runnable.run()
        }
    }
}
